### PR TITLE
PLT-7591: Improved Auditor Report Upload

### PR DESCRIPTION
**Pull request by EehMauro**
@coderabbitai: ignore

### CODE REVIEW

**Comment by amnambiar** [comment]
Note: we need to integrate changes from PR #8 here once that is approved/merged

**Comment by RSoulatIOHK** [comment]
Love the new design.

### Error messages
Ok, I don't know why, but at first, the checks for the values were not happening until I clicked on Send report. And after that it would only show me one field that was wrong. I fixed them in the order requested:

- Auditor Information
    - Name 
    - Website
    - Logo

And then, it showed all the other field with wrong values, and will display live the error messages of everything I filled wrong.

### Wrong Regex
I don't get from the code why this Website value (example.com or www.example.com) is not ok for the regex:
image

**Comment by EehMauro** [comment]
> Love the new design.
> 
> ### Error messages
> Ok, I don't know why, but at first, the checks for the values were not happening until I clicked on Send report. And after that it would only show me one field that was wrong. I fixed them in the order requested:
> 
> * Auditor Information
>   
>   * Name
>   * Website
>   * Logo
> 
> And then, it showed all the other field with wrong values, and will display live the error messages of everything I filled wrong.
> 
> ### Wrong Regex
> I don't get from the code why this Website value (example.com or [www.example.com](http://www.example.com)) is not ok for the regex: image

Fixed!

**Comment by EehMauro** [comment]
Added Auditor Report Upload fields prefill with user profile

**Comment by amnambiar** [comment]
@EehMauro , Here are my thoughts -
'Auditor Report Upload' and ['Review Certification Metadata'](PR #8) are same just with Subject and Report URLs missing in the later. Rest all schema, functionalities are the same. Hence I created a common MetadataForm component to share between the two; by supplying separate configurations to build the form, and schemas (see PR #20 for latest).

Have you kept this in mind in this PR, otherwise what is your suggestion. We need to discuss over this, to reduce redundancy and make this simpler.

**Comment by amnambiar** [comment]
@EehMauro , Also I see you are not making use of schema validators like yup. Is there a particular reason why? Also, do you advocate in removing them throughout the app?

**Comment by amnambiar** [comment]
> @EehMauro , Here are my thoughts - 'Auditor Report Upload' and ['Review Certification Metadata'](PR #8) are same just with Subject and Report URLs missing in the later. Rest all schema, functionalities are the same. Hence I created a common MetadataForm component to share between the two; by supplying separate configurations to build the form, and schemas (see PR #20 for latest).
> 
> Have you kept this in mind in this PR, otherwise what is your suggestion. We need to discuss over this, to reduce redundancy and make this simpler.

PR #8 was merged after a lot of integration testing / bug fixes / schema corrections :laughing: 
If we redo the component entirely in this PR , wouldn't we have to apply the same effort here as well? I'm a bit sceptical with this :upside_down_face:

**Comment by EehMauro** [comment]
Hi @amnambiar, I can merge #20 here so I can re-use the form you made and the yup validations

File: src/compositions/InputGroup/utils/schemas.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  .matches(/[0-9a-fA-F]{64}/, 'The field value must be a valid hash with 64 characters');
+
+export const CommitHash = yup.string()
+  .min(7, 'The field value must be at least 7 characters of the commit hash')
+  .max(40, 'The field value must has more than 40 characters')
+  .matches(/^[a-f0-9]{7,40}$/, 'The field value must be a valid commit hash');
+
+export const WalletAddress = yup.string()
+  .matches(/^(addr_test1|addr1)[a-zA-Z0-9]{53,}$/, {
+    message: 'The field value must be a valid wallet address',
```
Comment (RSoulatIOHK) [inline src/compositions/InputGroup/utils/schemas.ts]:
```suggestion
    message: 'The field value must be a valid contract address',
```

File: src/utils/utils.ts [hunk 2]
```diff
@@ -86,6 +86,11 @@ export const transformEmptyStringToNullInObj = (obj: any) => {
   return obj;
 }
 
+export const removeEmptyStringsDeep = (obj: any) => {
+  Object.keys(obj).forEach(key => (obj[key] && typeof obj[key] === 'object') && removeEmptyStringsDeep(obj[key]) || (obj[key] === '') && delete obj[key]);
```
Comment (RSoulatIOHK) [inline src/utils/utils.ts]: The logic is probably correct but it's kind of hard to follow, could you add some parantheses to better separate the `and`s and `or`s

File: src/pages/session/index.tsx [hunk 3]
```diff
@@ -27,7 +31,7 @@ export default () => {
     }
   }, [hasAnActiveSubscription, isSessionFetched, location.pathname]);
 
-  if (!isSessionFetched) {
+  if (!isSessionFetched || hasAnActiveSubscription && profile === null) {
```
Comment (RSoulatIOHK) [inline src/pages/session/index.tsx]: Could you add some parantheses to make it easier to follow the logic?

File: src/compositions/InputGroup/utils/index.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  return !required ? schema : schema.required('The field is required');
+};
+
+const buildSchema = (fields: Field[]): AnyObjectSchema => {
+  const schema = {} as {[key: string]: AnySchema};
+  for (const field of fields) {
+    if (field.type == FieldType.Object) {
+      schema[field.name] = buildSchema(field.fields || []);
+    } else if (field.type == FieldType.Array) {
```
Comment (RSoulatIOHK) [inline src/compositions/InputGroup/utils/index.ts]: I don't really know TS so sorry if this is a dumb question, but wouldn't it be better to use "==="?

File: src/compositions/InputGroup/utils/index.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    case FieldType.GitHubRepository: schema = Schema.GitHubRepository; break;
+    case FieldType.Dropdown: schema = Schema.Dropdown; break;
+  }
+  return !required ? schema : schema.required('The field is required');
+};
+
+const buildSchema = (fields: Field[]): AnyObjectSchema => {
+  const schema = {} as {[key: string]: AnySchema};
+  for (const field of fields) {
+    if (field.type == FieldType.Object) {
```
Comment (RSoulatIOHK) [inline src/compositions/InputGroup/utils/index.ts]: I don't really know TS so sorry if this is a dumb question, but wouldn't it be better to use "==="?

**Review by RSoulatIOHK — CHANGES_REQUESTED** [review]
Two more things but I don't know where to point it into the code but:

- The checks on the values don't start until I click a first time on "Send report"
- If I put a too long "Script Hash" like : `01234567890123456789012345678901234567890123456789012345678901234567890012345678901234567890123456789012345678901234567890123456789012345678900123456789012345678901234567890123456789012345678901234567890123456789001234567890123456789012345678901234567890123456789012345678901234567890` it never shows on the form that it is wrong but I get an error when clicking on "Send report"

File: src/compositions/InputGroup/utils/schemas.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import * as yup from "yup";
+
+export const Text = yup.string();
+
+export const Email = yup.string()
+  .email('The field value must be a valid email address');
+
+export const Subject = yup.string()
```
Comment (RSoulatIOHK) [inline src/compositions/InputGroup/utils/schemas.ts:8]: Subject is more complex than that:
"Subject: string
pattern: ^[A-Za-z0-9_]{1,64}$"

**Review by RSoulatIOHK — CHANGES_REQUESTED** [review]
Sorry, Missed this during the first review

**Comment by amnambiar** [comment]
1. reportURLs are to be sent as array of strings in report.reportURLs in the payload
2. Move Add Script to after all the the DApp Scripts (like Romain mentioned)
3. Error in form submit is retained while navigating away and coming back, even though the form is cleared. Clear errors in the form, while moving away from the page.

@EehMauro bugs/changes to be addressed here

**Comment by amnambiar** [comment]
@EehMauro , Pardon my bad. Can you please fix (1)

POST /auditor/reports expects . Can you fix that. And, the error `400 user error (Report URLs do not match)` when you give urls to different PDFs are because they are basically NOT different urls to the same report file(expected). 
```
"report": [
    "string"
  ],
```

**Comment by amnambiar** [comment]
@EehMauro Bugs - In Review Certification Metadata -- 
1. remove fields Certification Level, Subject, Report URLs
2. Add a cancel or close button to cancel the modal if he chooses to, as this is not mandatory

File: src/pages/certification/Certification.tsx [hunk 7]
```diff
@@ -94,13 +94,11 @@ const Certification = () => {
...
 
     return (
-        <div className="content-area">
-            <div className="content-area-title-section py-7 flex justify-between">
-                <h2 className="m-0">
-                    {runId
-                        ? "Run: " + ellipsizeString(runId)
-                        : "Testing Tool"}
-                </h2>
+        <Container className="py-4" maxWidth="xl">
```
Comment (amnambiar) [inline src/pages/certification/Certification.tsx]: @EehMauro ,

this causes a max-width to content-section which doesn't actually expand the content to fit the screen when >4k. We don't want <Container> elements for that reason. Atleast in Testing and History views. 

I am reverting this change here, as it also breaks the button styles (Test again, New test)

Otherwise make sure to use  maxWidth={false} like in FullReportTable.tsx

**Comment by amnambiar** [comment]
>  In Review Certification Metadata --
> 
> remove fields Certification Level, Subject, Report URLs
> Add a cancel or close button to cancel the modal if he chooses to, as this is not mandatory

This is done.

**Review by amnambiar — APPROVED** [review]
LGTM now

**Review by RSoulatIOHK — APPROVED** [review]
